### PR TITLE
R4R: implement meta transaction

### PR DESCRIPTION
**Pull request by HaoyangLiu**
This PR enable mantle-v2 to natively support meta transaction(gas less. transaction) in protocol. 
There are other gasless transaction protocols which don't need any changes in protocol, such as 
https://eips.ethereum.org/EIPS/eip-2771
https://eips.ethereum.org/EIPS/eip-2612
However, all above gas less transaction protocols have some limitations on either user accounts or assets contracts, which is not convenient.

Tools to build and broadcast meta tx:
https://github.com/HaoyangLiu/golang-tool/blob/main/cmd/metatx/main.go

### CODE REVIEW

File: core/txpool/txpool.go
Comment (wukongcheng) [inline core/txpool/txpool.go:1]: Can you add Unit Test to txpool?

File: core/txpool/txpool.go [hunk 1]
```diff
@@ -1443,6 +1461,35 @@ func (pool *TxPool) reset(oldHead, newHead *types.Header) {
...
+	for _, tx := range list.txs.Flatten() {
+		metaTxParams, err := types.DecodeAndVerifyMetaTxParams(tx)
+		if err != nil {
+			invalidMetaTxs = append(invalidMetaTxs, tx)
+			continue
+		}
+		if metaTxParams == nil {
+			continue
+		}
+		if metaTxParams.ExpireHeight < currHeight {
```
Comment (wukongcheng) [inline core/txpool/txpool.go:1478]: What is the function of st.msg. MetaTxParams. ExpireHeight?

File: core/txpool/txpool.go [hunk 2]
```diff
@@ -688,7 +688,21 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
...
+	metaTxParams, err := types.DecodeAndVerifyMetaTxParams(tx)
+	if err != nil {
+		return err
+	}
+	var balance *big.Int
+	if metaTxParams != nil {
+		if metaTxParams.ExpireHeight < pool.chain.CurrentBlock().Number.Uint64() {
+			return types.ErrExpiredMetaTx
+		}
+		balance = pool.currentState.GetBalance(metaTxParams.GasFeeSponsor)
```
Comment (wukongcheng) [inline core/txpool/txpool.go:701]: What happens when there is not enough balance in st.msg. MetaTxParams. GasFeeSponsor

Comment (HaoyangLiu) [reply core/txpool/txpool.go:701]: In line 706, `balance` will be checked against tx cost.

File: core/txpool/txpool.go (hunk 1 above)
Comment (HaoyangLiu) [reply core/txpool/txpool.go:1478]: The gas fee sponsor have no idea about when the users will submit meta transactions. Suppose the gas fee sponsor don't want to sponsor gas fee any more, and want to use the wallet in the future, it would be possible that sponsor wallet will be deducted native token in the future, which will be very confusing. With the `ExpireHeight`,  the sponsor can ensure that there will be no unconscious deduction after the highest `ExpireHeight`. 
In summary, the `ExpireHeight` can bring more determinism for the fee sponsor.

File: core/txpool/txpool.go
Comment (HaoyangLiu) [reply core/txpool/txpool.go:1]: Good idea.

**Comment by HaoyangLiu** [comment]
Try to support gas fee sponsor percent.

**Review by byteflyfunny — APPROVED** [review]

File: core/state_transition.go [hunk 3]
```diff
@@ -241,6 +248,9 @@ func (st *StateTransition) to() common.Address {
 }
 
 func (st *StateTransition) buyGas() error {
+	if err := st.applyMetaTransaction(); err != nil {
+		return err
+	}
 	mgval := new(big.Int).SetUint64(st.msg.GasLimit)
```
Comment (Tri-stone) [inline core/state_transition.go:254]: `st.msg.GasLimit` is set by user and is bigger than the actual `gasused`,  in [refundGas](https://github.com/mantlenetworkio/op-geth/blob/a33e4e11416873c14d868c537afe7e660936ba8f/core/state_transition.go#L519) the remaining gas will refund back to `msg.sender` instead of `sponsor`. is it possible that `msg.sender` steals asset from `sponsor`?

File: core/state_transition.go [hunk 4]
```diff
@@ -526,7 +571,16 @@ func (st *StateTransition) refundGas(refundQuotient uint64) {
 
 	// Return ETH for remaining gas, exchanged at the original rate.
```
Comment (Tri-stone) [inline core/state_transition.go:572]: In EVM there is refund mechanism that if evm storage setting from 1 to 0, the gas will refund to `msg.sender`.
This refund gas will be distributed to `msg.sender` and `sponsor` now.

**Review by Tri-stone — APPROVED** [review]
LGTM
